packages/grafana-data/src/types/datasource.ts

```diff
@@ -5,18 +5,32 @@ import { PanelData } from './panel';
 import { LogRowModel } from './logs';
 import { AnnotationEvent, KeyValue, LoadingState, TableData, TimeSeries } from './data';
 import { DataFrame, DataFrameDTO } from './dataFrame';
-import { RawTimeRange, TimeRange } from './time';
+import { RawTimeRange, TimeRange, AbsoluteTimeRange } from './time';
 import { ScopedVars } from './ScopedVars';
 
 export interface DataSourcePluginOptionsEditorProps<JSONData = DataSourceJsonData, SecureJSONData = {}> {
   options: DataSourceSettings<JSONData, SecureJSONData>;
   onOptionsChange: (options: DataSourceSettings<JSONData, SecureJSONData>) => void;
 }
 
+export type DataSourceQuery<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
+  infer TQuery,
+  infer _TOptions
+>
+  ? TQuery
+  : never;
+
+export type DataSourceOptions<DSType extends DataSourceApi<any, any>> = DSType extends DataSourceApi<
+  infer _TQuery,
+  infer TOptions
+>
+  ? TOptions
+  : never;
+
```

Comment:
**head... exploding**

I'll trust you on this one :)  The nested typing on the datasources getting super hard to follow - I think that is just the nature of the structure, but is something we should try to clean up if possible

Reply:
Agreed